### PR TITLE
feat: Add logic to export and import etcd catalog keys as JSON

### LINKED ISSUE #176: ice cli command to export/import catalog data
New ice cli commands to
`export-catalog` - The catalog data from etcd can be exported as json.
`import-catalog` - Import from file .

### PR DESCRIPTION
**Pull request by subkanthi**
…ON to catalog.

closes: #https://github.com/Altinity/ice/issues/27
closes: https://github.com/Altinity/ice/issues/176

### CODE REVIEW

**Comment by subkanthi** [comment]
## Export
```
ice catalog-export -o catalog_export.json
2026-05-21 15:23:20 [49717579-57/47868] INFO c.a.i.r.c.i.r.CatalogAdminServlet > Exported 1 namespace(s) and 13 table(s)
2026-05-21 15:23:20 [main/74519] INFO com.altinity.ice.cli.Main > Wrote catalog snapshot to catalog_export.json
```

## Delete all keys
```
etcdctl del --prefix ""
```

## Import

```
ice catalog-import -i catalog_export.json          
2026-05-21 15:25:25 [49717579-38/47868] INFO c.a.i.r.c.i.r.CatalogAdminServlet > Import: 14 created, 0 skipped, 0 overwritten
{
  "created" : 14,
  "skipped" : 0,
  "overwritten" : 0,
  "catalog_name" : "default",
  "exported_at" : "2026-05-21T20:23:20.555925Z"
}
```

```
ice list-tables ontime            
2026-05-21 15:26:57 [49717579-41/47868] INFO c.a.i.r.c.i.r.RESTCatalogServlet > @token:anonymous GET v1/config
2026-05-21 15:26:57 [49717579-42/47868] INFO c.a.i.r.c.i.r.RESTCatalogServlet > @token:anonymous GET v1/namespaces/ontime/tables
---
namespace: ontime
tables:
- ontime.testmay6_17_01
- ontime.testmay6_19_01
- ontime.testmay7_13_01
- ontime.testmay7_15_01
- ontime.testmay7_16_01
- ontime.testmay7_16_02
- ontime.testmay7_17_02
- ontime.testmay7_17_03
- ontime.testmay7_17_04
- ontime.testmay7_17_05
- ontime.testmay7_17_06
- ontime.testmay7_17_07
- ontime.testmay9_17_07
```

File: ice-rest-catalog/src/main/java/com/altinity/ice/rest/catalog/internal/cmd/CatalogAdminService.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    for (CatalogSnapshot.NamespaceEntry entry : snapshot.namespaces()) {
+      EtcdCatalog.PutCatalogKvResult result =
+          etcdCatalog.putCatalogKv(entry.key(), marshal(entry.value()), overwrite, dryRun);
+      switch (result) {
+        case CREATED -> created++;
+        case SKIPPED -> skipped++;
+        case OVERWRITTEN -> overwritten++;
+      }
+    }
+    for (CatalogSnapshot.TableEntry entry : snapshot.tables()) {
```
Comment (xieandrew) [inline ice-rest-catalog/src/main/java/com/altinity/ice/rest/catalog/internal/cmd/CatalogAdminService.java]: Both for loops will cause a NullPointerException if there is no `namespaces` or `tables` key in the JSON file. Ex: `org.apache.iceberg.exceptions.RESTException: Unhandled error: ErrorResponse(code=500, type=null, message={"error":{"message":"Cannot invoke \"java.util.List.iterator()\" because the return value of \"com.altinity.ice.rest.catalog.internal.cmd.CatalogSnapshot.tables()\" is null","type":"NullPointerException","code":500}})`

File: ice-rest-catalog/src/main/java/com/altinity/ice/rest/catalog/Main.java [hunk 2]
```diff
@@ -337,6 +338,9 @@ private static Server createBaseServer(
     var h = new ServletHolder(new RESTCatalogServlet(restCatalogAdapter));
     mux.addServlet(h, "/*");
 
+    var adminServlet = new ServletHolder(new CatalogAdminServlet(catalog, config.name()));
```
Comment (xieandrew) [inline ice-rest-catalog/src/main/java/com/altinity/ice/rest/catalog/Main.java]: This method (createBaseServer) is called in both createServer and createAdminServer, so the admin servlet seems accessible to anyone with ice access. Is that intentional?

File: ice/src/main/java/com/altinity/ice/cli/Main.java [hunk 3]
```diff
@@ -813,6 +815,88 @@ void deleteNamespace(
...
+      @CommandLine.Option(
+              names = {"-o", "--output"},
+              description = "Output file path (- for stdout, default)")
+          String output,
+      @CommandLine.Option(
+              names = "--namespace",
+              description = "Export only this namespace and its tables (e.g. flowers)")
+          String namespace)
+      throws IOException {
+    try (CatalogAdminClient client = createCatalogAdminClient(this.configFile())) {
```
Comment (xieandrew) [inline ice/src/main/java/com/altinity/ice/cli/Main.java:832]: The implementation of this command (and its helper method) should be in a different class (not Main) in internal/cmd like the other commands do.

File: ice/src/main/java/com/altinity/ice/cli/Main.java [hunk 4]
```diff
@@ -813,6 +815,88 @@ void deleteNamespace(
...
+              description = "Input file path (- for stdin, default)")
+          String input,
+      @CommandLine.Option(names = "--dry-run", description = "Preview changes without writing")
+          boolean dryRun,
+      @CommandLine.Option(
+              names = "--overwrite",
+              description = "Replace existing keys (default: skip existing keys)")
+          boolean overwrite)
+      throws IOException {
+    String snapshotJson = readCatalogSnapshotInput(input);
```
Comment (xieandrew) [inline ice/src/main/java/com/altinity/ice/cli/Main.java]: Same thing as catalogExport, the implementation should be in a different class.

**Review by xieandrew — CHANGES_REQUESTED** [review]
Functionality looks good, tested locally with 1 and 3 node etcd using docker compose. I have a couple comments on a potential security issue with the endpoints and an edge case error.

Also, it would be nice if there was a test in EtcdCatalogIT for this, but we could add it in a later PR.

**Review by xieandrew — APPROVED** [review]
Looks good to me
